### PR TITLE
Gone as far as I can with rails g controller and models. Having trouble accessing db.

### PR DESCRIPTION
**Pull request by mlworle**
Some questions:
1. How is create action different from new action? as in student_controller def create, which I added, and def new, which looks empty.
2. How do we access the database? I think I can do it from the command line, but that's not the point. I want the user to do it.
3. Followup on 2: So is new.html.erb suppose to have a link to the new or create action? If so, is this a javascript button? How does that work?
4. What is the purpose of the show action? This is just another def written in student_controller.rb, right?
5. Cosmetics: since the root is now index.html.erb, how do I get a link to new.html.erb and start the process of entering data?

I'm spinning in code and could use some direction.
Thanks!

### CODE REVIEW

File: app/views/student/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<h1> Attinderance | Student Index </h1>
+
+<p>This is where a list of all students will be found. Access to the database needs to be established. Still trying to figure that out :(</P
```
Comment (edwardloveall) [inline app/views/student/index.html.erb]: I think you forgot to close the paragraph tag here.

**Comment by edwardloveall** [comment]
>1. How is create action different from new action? as in student_controller def create, which I added, and def new, which looks empty.
`create` and `new` are similar in purpose which might be part of your confusion. `new` brings the user to a page to fill out information about what will be created. `create` takes that information and actually saves it in a database. It isn't a page that the users ever sees (for the most part), but rather how the information gets created in the database. Maybe it would help to think about it like:

* `new` - Filling out a form for a drivers license at the DMV
* `create` - The DMV processes your form and returns to you that drivers license (or tells you to go fill it out again if you made an error)

**Comment by edwardloveall** [comment]
>2. How do we access the database? I think I can do it from the command line, but that's not the point. I want the user to do it.

It's a multifaceted question, but I think you are asking "How do I interact with database rows with Ruby on Rails?" The answer is ActiveRecord which lets you do things like:

* `Student.new` to create a new, empty instance of a student
* `Student.create({ name: 'Mara Worle' })` to insert a new student row into your database with a name.
* `student.find_by({ name: 'Edward Loveall' })` to retrieve one student from the database with a specific name.

There are many more things you can do too. The next things you might need are `update`, and `destroy`.

**Comment by edwardloveall** [comment]
>3. Followup on 2: So is new.html.erb suppose to have a link to the new or create action? If so, is this a javascript button? How does that work?

No need to link to the new/create action from the new page in this case. The user needs to get to the new page somehow, but once they are on that page, they don't need another link to that same page.

To get to a page, we use what's called **link helpers** or **path helpers**. Here's how they work: When you type `rake routes` in your terminal, you get a list like:

```
      Prefix Verb   URI Pattern                  Controller#Action
        root GET    /                            student#index
    students GET    /students(.:format)          students#index
             POST   /students(.:format)          students#create
 new_student GET    /students/new(.:format)      students#new
edit_student GET    /students/:id/edit(.:format) students#edit
     student GET    /students/:id(.:format)      students#show
             PATCH  /students/:id(.:format)      students#update
             PUT    /students/:id(.:format)      students#update
             DELETE /students/:id(.:format)      students#destroy
```

The `Prefix` column is the name of the helper. When you are in a view or controller, you can enter something like `students_path` or `students_url`; that is, the prefix, plus either `_path` or `_url`. These will yield different outputs. For example, to get to the page to create a new student, you would use something like

```erb
<%= link_to 'New Student', new_student_path %>
```

In a view. That will make some html like:

```html
<a href="/students/new">New Student</a>
```

Here's documentation if you're curious: http://guides.rubyonrails.org/routing.html

**Comment by edwardloveall** [comment]
>4. What is the purpose of the show action? This is just another def written in student_controller.rb, right?

It is written in `students_controller.rb`, yes. (technically, if it's an "action" it's inside of a controller. otherwise it's just called a method)

It's purpose is to show information about some resource, like, say a student. Think of it at someone's profile page that anyone can get to. Like for example if I go to `https://twitter.com/edwardloveall`, that's Twitter's version of `users#show`. This very page, in fact, is a show page. It's showing one particular pull request instead of a group of pull requests, or creating a new pull request, or updating, etc.

**Comment by edwardloveall** [comment]
>5 Cosmetics: since the root is now index.html.erb, how do I get a link to new.html.erb and start the process of entering data?

Since `root` is pointed at `students#index`, you'll need to put a new link (like above in the answer to question 3) on the `index.html.erb` page. Clicking that link will move away from the index page and load the new page.

File: config/routes.rb [hunk 2]
```diff
@@ -1,11 +1,17 @@
 Rails.application.routes.draw do
+  root 'student#index'
+
+  get 'student/new'
```
Comment (edwardloveall) [inline config/routes.rb]: These routes seem wrong to me. The way `get` works is you say what url you're looking for, then, when the system gets that url, where should it go? You haven't told it where to go. But that all is taken care of for you if you use:

```ruby
resources :students
```

This makes all the routes you need for index, show, new, create, edit, update, and delete. I'd highly recommend using that instead. Then you can get rid of all these get's in here.

**Comment by edwardloveall** [comment]
Good work here. Sorry for the wall of text. I added some specific comments on lines of code. You can see them if you click on **Files changes** at the top.

Next steps would be to get your controller actions filled with data, and your views using that data. Tricky step. The `show` action is probably the easiest if you want to use it as a test. `new` or `index` would be next after that I think in terms of complexity. Do whatever order you like though. Technically this branch is just for creating students :smile:

**Comment by edwardloveall** [comment]
Looking good! Remember that extra `student` (instead of `students`) directory? We need to get rid of it. From the Attinderence directory run:

```shell
git rm -r app/views/student
```

**Comment by edwardloveall** [comment]
After that, run through the app and do a final check to see if it does what you expect. Visit the new page, create a new student, visit the index, check to see if that student is there, etc. Anything you can think of that you worked on in this branch.

If everything looks good leave a comment here.

**Comment by mlworle** [comment]
Two things are buggy.
1. after a student is created, the new page is empty. I'd like to either redirect to the index or have link to redirect. How to do that DRYly?
2. How to delete/destroy the data in ActiveRecord to clean up the students inputted for testing. Rails console is not happy.

**Comment by mlworle** [comment]
Oh, bugger.
I committed suggestotron to this branch. Have not pushed. Is is possible to uncommit?

**Comment by edwardloveall** [comment]
Run `git log`. You'll probably see something like:

```
commit 8d417c93a1e15a95f64b7ad7f47b2776d8fe0a82
Author: mlworle
Date:   Mon Jul 27 14:53:24 2015 -0400

    Bad commit

commit 6bf72ec3daf4d1cb79e64b620fc49ebc0ced80e7
Author: mlworle
Date:   Tue Jul 21 13:35:18 2015 -0400

    Bad commit

commit e70c6d1e85403ab244e67b2f0fddca2b66d0f91d
Author: mlworle
Date:   Mon Jul 20 11:08:40 2015 -0400

    Good commit

...
```

You want that commit number from that good commit just before the bad ones. Copy that.

Then run `git reset --hard <good-commit-number>` eg `git reset --hard e70c6d1e85403ab244e67b2f0fddca2b66d0f91d` in this case.

Finally, run `git status`. If you see any files that are untracked (i.e. not modified or deleted) remove them by hand. That should work.

**Comment by edwardloveall** [comment]
Your other questions:

>after a student is created, the new page is empty. I'd like to either redirect to the index or have link to redirect. How to do that DRYly?

In the `StudentController#create` action:

```ruby
def create
  @student = Student.new(student_params)

  @student.save
  redirect_to @student
end
```

that last `redirect_to` line does exactly what it says: redirects to somewhere. If you give it an instance of a model like `@student`, it's a shorthand for `student_path(@student)` which renders the currently blank show page. If you want to redirect to the index, you could use: `students_path`.

>How to delete/destroy the data in ActiveRecord to clean up the students inputted for testing. Rails console is not happy.

If you really want to delete all students, in the rails console, try `Student.destroy_all`. This will remove *all* students, even legitimate ones, so make sure it's what you want to do. Let me know if you're looking for a more nuanced approach.

**Comment by mlworle** [comment]
Actually, the commits were all good on this branch. Good to know for future. Database clear.
Now what?

**Comment by edwardloveall** [comment]
Last step; merge.

Switch to your master branch with `git checkout master`

Then merge in this branch with `git merge mw-create-students`

Finally, push all the changes up to master: `git push origin master`

**Comment by mlworle** [comment]
Objections to rebasing first?
Must be in application directory to git rebase --interactive master?

**Comment by edwardloveall** [comment]
Yeah you can do that. I was trying to not overcomplicate things, but yes. Interactive rebasing is a great idea.

Make sure you're on the `mw-create-students` branch and then run your command.

**Pull request merged**